lib/SILGen/SILGenApply.cpp

```diff
@@ -4474,8 +4496,10 @@ RValue SILGenFunction::emitApply(ResultPlanPtr &&resultPlan,
   // hop back to the current executor
   breadcrumb.emit(*this, loc);
 
-  // Pop the argument scope.
-  argScope.pop();
+  // For non foreign async function calls, emit cleanup here.
+  if (!calleeTypeInfo.foreign.async) {
+    argScope.pop();
+  }
```

Comment:
Hmmm. I was delaying the `argScope.pop()` in order to not generate `destroy_value` or other cleanups of the args before `await_async_continuation` which is generated after the call.

Reply:
@jckarter I was just talking with @meg-gupta and we realized the following things:

1. Most likely the /real/ solution here is to not have the result planner insert the await_async_continuation. That being said, we don't have time to do that now = p. We should probably file a radar or SR to track fixing this.

2. Is there any issue if we insert an extra copy and lifetime extend in this specific case these arguments over the result planner. We would do it by hand without relying on a scope pop. To be more specific:

a. Before the result planner, in this case we still do argScope.pop() always, but before we do that, we create a SmallVector<ManagedValue>, do the loop from 4581 and insert a copy value using the SILValue SILBuilder API and then stash these new SILValues into the SmallVector<ManagedValue> using ManagedValue::forUnmanaged. 
b. Right before the result planner, we loop over the SmallVector again, but this time, create the relevant cleanup (which will insert a fix_lifetime + destroy_Value). This will ensure that if we hit a throwing case, we insert the cleanup along the throwing path.
c. After the result planner, we go through the ManagedValue small vector again and insert the fix lifetime/destroy by hand and forward the ManagedValue cleanups.

This I think preserves the current scoping structure and makes this a very tailored fix that doesn't have the possibility of emergent effects that may result from moving where scopes are.

@jckarter your thoughts?

Reply:
Thanks @gottesmm for taking a look quickly and your suggestions! SemanticARCOpts/MandatoryARCOpts get rid of the copy.

Reply:
Yes. That was my thought. This is a case where we can do something a bit simpler and rely on the optimizer to do a good enough job.

Reply:
That sounds good. My concern was more about spurious exclusivity failures because borrows got overextended, but if you don't think that's an issue then I'm OK with it.

Reply:
I don't think that will happen since we are making a copy of the value. So it should be independent of any exclusivity scopes or the like.